// libxmljs/dist/lib/version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeCount = exports.memoryUsage = exports.libxml_debug_enabled = exports.libxml_parser_version = exports.libxml_version = exports.version = void 0;
var version = require("../../package.json").version;
exports.version = version;
var functions_1 = require("./bindings/functions");
Object.defineProperty(exports, "memoryUsage", { enumerable: true, get: function () { return functions_1.getMemUsed; } });
Object.defineProperty(exports, "nodeCount", { enumerable: true, get: function () { return functions_1.getNodeCount; } });
var constants_1 = require("./bindings/constants");
var variables_1 = require("./bindings/variables");
var libxml_version = constants_1.VERSION;
exports.libxml_version = libxml_version;
var libxml_parser_version = (0, functions_1.__xmlParserVersion)();
exports.libxml_parser_version = libxml_parser_version;
var libxml_debug_enabled = !!variables_1.libxmljs_debug;
exports.libxml_debug_enabled = libxml_debug_enabled;
